src/vs/workbench/services/semanticSimilarity/common/semanticSimilarityService.ts

```diff
@@ -5,25 +5,118 @@
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
 
-export const ISemanticSimilarityService = createDecorator<ISemanticSimilarityService>('IEmbeddingsService');
+export const ISemanticSimilarityService = createDecorator<ISemanticSimilarityService>('ISemanticSimilarityService');
 
 export interface ISemanticSimilarityService {
+	readonly _serviceBrand: undefined;
+
 	isEnabled(): boolean;
 	getSimilarityScore(string1: string, comparisons: string[], token: CancellationToken): Promise<number[]>;
+	registerSemanticSimilarityProvider(provider: ISemanticSimilarityProvider): IDisposable;
+	unregisterSemanticSimilarityProvider(provider: ISemanticSimilarityProvider): void;
+}
+
+export interface ISemanticSimilarityProvider {
+	provideSimilarityScore(string1: string, comparisons: string[], token: CancellationToken): Promise<number[]>;
 }
 
+export class SemanticSimilarityService implements ISemanticSimilarityService {
+	readonly _serviceBrand: undefined;
+
+	static readonly DEFAULT_TIMEOUT = 1000 * 10; // 30 seconds
+
+	private readonly _providers: ISemanticSimilarityProvider[] = [];
+	// remove when we move over to API
+	private readonly oldService: OldSemanticSimilarityService;
+
+	constructor(
+		// for the old service
+		@ICommandService commandService: ICommandService,
+		@IFileService fileService: IFileService
+	) {
+		this.oldService = new OldSemanticSimilarityService(commandService, fileService);
+	}
+
+	isEnabled(): boolean {
+		return this._providers.length > 0;
+	}
+
+	registerSemanticSimilarityProvider(provider: ISemanticSimilarityProvider): IDisposable {
+		this._providers.push(provider);
+		return {
+			dispose: () => {
+				const index = this._providers.indexOf(provider);
+				if (index >= 0) {
+					this._providers.splice(index, 1);
+				}
+			}
+		};
+	}
+
+	unregisterSemanticSimilarityProvider(provider: ISemanticSimilarityProvider): void {
+		const index = this._providers.indexOf(provider);
+		if (index >= 0) {
+			this._providers.splice(index, 1);
+		}
+	}
+
+	async getSimilarityScore(string1: string, comparisons: string[], token: CancellationToken): Promise<number[]> {
+		if (this._providers.length === 0) {
+			// Remove when we have a provider shipping in extensions
+			if (this.oldService.isEnabled()) {
+				return this.oldService.getSimilarityScore(string1, comparisons, token);
+			}
+			throw new Error('No semantic similarity providers registered');
+		}
+
+		const cancellablePromises: Array<CancelablePromise<number[]>> = [];
+
+		const timer = timeout(SemanticSimilarityService.DEFAULT_TIMEOUT);
+		const disposible = token.onCancellationRequested(() => {
+			disposible.dispose();
+			timer.cancel();
+		});
+
+		for (const provider of this._providers) {
+			cancellablePromises.push(createCancelablePromise(async t => {
+				try {
+					return await provider.provideSimilarityScore(string1, comparisons, t);
+				} catch (e) {
+					// logged in extension host
+				}
+				await timer;
```

Comment:
This code won't run when it's supposed to time out, it will be blocked by awaiting the call to the provider

Reply:
Or I guess that call throws when the whole set is canceled due to the timer promise throwing?

Reply:
This is actually my way of tricking `raceCancelablePromises` into waiting for the first promise that _resolves_ not rejects. If one of the providers reject, then we catch the error and wait until the timer is done... this allows _another provider_ time to resolve correctly and everything else will be cancelled when either:
* something in the array resolves
* the DEFAULT_TIMER resolves

Reply:
I'll add a comment in a follow up